# visualize coverage of 10x libraries

# libraries:

# Illumina sequencing:
# Pool116_3-6: 10x 5' scRNA-seq
#
# whole-transcriptome sequencing with ONT:
# 3005: 10x 5' scRNA-seq with clean-up PCR1
# 3HTcDNA1: 10x 3' high-throughput (HT) scRNA-seq without clean-up PCR1
# 3HTcDNA2: 10x 3' high-throughput (HT) scRNA-seq with clean-up PCR1
# 3HTcDNA2_bc: 10x 3' high-throughput (HT) scRNA-seq with clean-up PCR1 and considering only reads with cell barcodes
#
# coverage was generated by aligning reads against a transcriptome reference 
# available under ./data/reference/gencode.v41.pc_canon_edit.fa.gz
# followed by bedtools genomecov -d -ibam $bamfile 

library(dplyr)
library(ggplot2)

samples = c('Pool116_3' = './data/benchmarking/WT_bulk_coverage/Pool116_3_pc_canon_coverage.txt.gz',
            'Pool116_4' = './data/benchmarking/WT_bulk_coverage/Pool116_4_pc_canon_coverage.txt.gz',
            'Pool116_5' = './data/benchmarking/WT_bulk_coverage/Pool116_5_pc_canon_coverage.txt.gz',
            'Pool116_6' = './data/benchmarking/WT_bulk_coverage/Pool116_6_pc_canon_coverage.txt.gz',
            '3005' = './data/benchmarking/WT_bulk_coverage/3005_1_WT_tx_coverage.txt.gz',
            '3HTcDNA1' = './data/benchmarking/WT_bulk_coverage/3HTcDNA1_tx.coverage.csv.gz',
            '3HTcDNA2' = './data/benchmarking/WT_bulk_coverage/3HTcDNA2_tx.coverage.csv.gz',
            '3HTcDNA2_bc' = './data/benchmarking/WT_bulk_coverage/3pWT_trns_deconcat.coverage.csv.gz')

for (s in seq(5,7)) {
  message(samples[s])
  coverage.df = data.table::fread(samples[s])
  colnames(coverage.df) = c('transcript', 'position', 'coverage')
  coverage.df$gene = stringr::str_split_fixed(coverage.df$transcript, pattern = '-', n=2)[,1]
  coverage.df$length = as.numeric(stringr::str_split_fixed(coverage.df$transcript, pattern = '_', n=2)[,2])
  
  coverage.df = coverage.df %>% group_by(transcript) %>% mutate(max.coverage = max(coverage))
  coverage.df = coverage.df %>% group_by(transcript) %>% mutate(int.coverage = sum(coverage) / length(transcript))
  coverage.df$coverage.bin = 'high'
  coverage.df$coverage.bin[which(coverage.df$max.coverage < 1000)] = 'med'
  coverage.df$coverage.bin[which(coverage.df$max.coverage < 200)] = 'low'
  coverage.df$length.bin = 'high'
  coverage.df$length.bin[which(coverage.df$length < 10000)] = 'med'
  coverage.df$length.bin[which(coverage.df$length < 5000)] = 'low'
  
  coverage.df$rel.coverage = coverage.df$coverage / coverage.df$max.coverage
  
  # exclude genes with minimal coverage
  coverage.df = coverage.df[-which(coverage.df$max.coverage < 50),]
  
  coverage.position = coverage.df %>% group_by(position, length.bin) %>% summarize(mean.coverage = mean(rel.coverage),
                                                                                   abs.coverage = mean(coverage),
                                                                                   transcripts = length(unique(transcript)))
  coverage.position = coverage.position %>% group_by(length.bin) %>% mutate(max.transcripts = max(transcripts))
  coverage.position$rel.transcripts = coverage.position$transcripts / coverage.position$max.transcripts
  
  
  p=ggplot() + 
    ggrastr::rasterize(geom_point(data=coverage.position, aes(x=position, y=abs.coverage, color=length.bin), size=0.5), dpi=600) + 
    #geom_point(data=coverage.position, aes(x=position, y=abs.coverage, color=length.bin), size=0.5) + 
    scale_x_sqrt(limits = c(1,20000), breaks = c(50,500,2000,5000,10000, 20000)) + 
    scale_y_continuous('coverage') +
    scale_color_manual(values = c('low' = '#4D6910', 'med' = '#FFA600', 'high' = '#FF1700')) + 
    theme_classic() + 
    theme(legend.position = 'none',
          axis.text = element_text('Arial', size=10, color='black'),
          axis.title = element_text('Arial', size=10, color='black')) 
  ggsave(paste0('./benchmarking/figures/WT_bulk_coverage/20220908_absolute_coverage_', names(samples)[s], '.svg'), width = 3, height = 2, plot = p)
  
  
  p=ggplot() + 
    ggrastr::rasterize(geom_point(data=coverage.position, aes(x=position, y=abs.coverage, color=length.bin), size=0.5), dpi=600) + 
    #geom_line(data=coverage.position, aes(x=position, y=abs.coverage, color=length.bin), size=0.5) + 
    scale_x_sqrt(limits = c(1,20000), breaks = c(50,500,2000,5000,10000, 20000)) + 
    scale_y_continuous('coverage') +
    scale_color_manual(values = c('low' = '#4D6910', 'med' = '#FFA600', 'high' = '#FF1700')) + 
    theme_classic() + 
    theme(legend.position = 'none',
          axis.text = element_text('Arial', size=10, color='black'),
          axis.title = element_text('Arial', size=10, color='black')) 
  ggsave(paste0('./benchmarking/figures/WT_bulk_coverage/20220908_absolute_coverage_', names(samples)[s], '.svg'), width = 3, height = 2, plot = p)
  
  ### plot from 3' end
  coverage.df$position_3p = coverage.df$length - coverage.df$position 
  coverage.position = coverage.df %>% group_by(position_3p, length.bin) %>% summarize(mean.coverage = mean(rel.coverage),
                                                                                   abs.coverage = mean(coverage),
                                                                                   transcripts = length(unique(transcript)))
  coverage.position = coverage.position %>% group_by(length.bin) %>% mutate(max.transcripts = max(transcripts))
  coverage.position$rel.transcripts = coverage.position$transcripts / coverage.position$max.transcripts
  
  p=ggplot() + 
    #ggrastr::rasterize(geom_point(data=coverage.position, aes(x=position_3p, y=abs.coverage, color=length.bin), size=0.5), dpi=600) + 
    geom_line(data=coverage.position, aes(x=position_3p, y=abs.coverage, color=length.bin), size=0.5) + 
    scale_x_sqrt(limits = c(1,20000), breaks = c(50,500,2000,5000,10000, 20000)) + 
    scale_y_continuous('coverage') +
    scale_color_manual(values = c('low' = '#4D6910', 'med' = '#FFA600', 'high' = '#FF1700')) + 
    theme_classic() + 
    theme(legend.position = 'none',
          axis.text = element_text('Arial', size=10, color='black'),
          axis.title = element_text('Arial', size=10, color='black')) 
  ggsave(paste0('./benchmarking/figures/WT_bulk_coverage/20221101_absolute_coverage_3p_', names(samples)[s], '.svg'), width = 3, height = 2, plot = p)
  
}

# coverage as function of transcript length
for (s in seq(1,6)) {
  message(samples[s])
  coverage.df = data.table::fread(samples[s])
  colnames(coverage.df) = c('transcript', 'position', 'coverage')
  coverage.df$gene = stringr::str_split_fixed(coverage.df$transcript, pattern = '-', n=2)[,1]
  coverage.df$length = as.numeric(stringr::str_split_fixed(coverage.df$transcript, pattern = '_', n=2)[,2])
  
  coverage.df.condensed = coverage.df %>% group_by(transcript) %>% summarize(transcript = unique(transcript), 
                                                                             length = unique(length),
                                                                             coverage.mean = mean(coverage))
  
  p=ggplot(coverage.df.condensed, aes(x=length, y=coverage.mean)) + 
    ggrastr::rasterize(ggpointdensity::geom_pointdensity(size=0.5), dpi=600) +
    scale_x_continuous('transcript length',limits = c(0,20000)) +
    scale_y_sqrt('unamplified cDNA coverage',breaks = c(1,10,50,250,500,1000,2500,5000, 10000), limits = c(0,10000)) + 
    scale_color_viridis_c() + 
    theme_classic() +
    theme(legend.position = 'bottom',
          axis.text = element_text('Arial', size=8, color='black'),
          axis.title = element_text('Arial', size=10, color='black')) 
  ggsave(paste0('./benchmarking/figures/WT_bulk_coverage/20221024_absolute_coverage_length_', names(samples)[s], '.svg'), width = 2, height = 2.5, plot = p)
}

### heatmaps
library(ComplexHeatmap)

for (s in seq(5,8)) {
  message(samples[s])
  coverage.df = data.table::fread(samples[s])
  colnames(coverage.df) = c('transcript', 'position', 'coverage')
  
  coverage.df$gene = stringr::str_split_fixed(coverage.df$transcript, pattern = '-', n=2)[,1]
  coverage.df$length = as.numeric(stringr::str_split_fixed(coverage.df$transcript, pattern = '_', n=2)[,2])
  coverage.df$position.rel = coverage.df$position / coverage.df$length
  
    
  coverage.df = coverage.df %>% group_by(transcript) %>% mutate(max.coverage = max(coverage))  
  coverage.df$coverage.rel = coverage.df$coverage / coverage.df$max.coverage
  coverage.df = coverage.df %>% mutate(position_bin = ntile(position.rel, n=100))
  
  coverage.df.rel = coverage.df %>% group_by(transcript) %>% filter(coverage != 0) %>% 
    group_by(transcript, position_bin) %>% 
    summarize(transcript = unique(transcript), 
              length = unique(length),
              coverage.mean = mean(coverage.rel))
  
  coverage.df.rel = coverage.df.rel %>% tidyr::pivot_wider(names_from = position_bin, values_from = coverage.mean)
  
  coverage.df.rel = coverage.df.rel[-which(duplicated(coverage.df.rel$length)),]
  coverage.df.rel = coverage.df.rel[order(coverage.df.rel$length),]
  
  mat = as.data.frame(coverage.df.rel[which(!is.na(coverage.df.rel$length)),as.character(seq(1,100))])
  rownames(mat) = as.numeric(coverage.df.rel$length[which(!is.na(coverage.df.rel$length))])
  
  subset = seq(1,nrow(coverage.df.rel), 250)
  labels = rownames(mat)[subset]
  
  col_fun = circlize::colorRamp2(breaks = seq(0,1,1/8), colors = BuenColors::jdb_palette(name = 'solar_extra'))
  svglite::svglite(paste0('./benchmarking/figures/WT_bulk_coverage/20221101_coverage_heatmap_', names(samples)[s], '.svg'), width = 4, height = 4)
  Heatmap(mat, cluster_columns = F, cluster_rows = F, show_row_names = F, show_column_names = F, col = col_fun, use_raster = T, raster_quality = 10) + 
    rowAnnotation(mark = anno_mark(at = subset, side = 'left', labels = labels, which = 'row', labels_gp = gpar(fontsize=8)))
  dev.off()
}
