# process amplicon data of comparison GoT versus nanoranger

# the experiment contains 5 conditions:

# GoT sequenced with Illumina (GoT.Illumina)
# GoT sequenced with ONT (GoT.ONT)
# nanoranger with normal cDNA (normal)
# nanoranger with cDNA generated by GoT (spikein)
# nanoranger with cDNA generated by GoT plus additive primer during PCR1 (spikein.additive)

library(dplyr)
library(ggplot2)
library(nanoranger.R)

AML1022 <- readRDS("./data/benchmarking/objects/20230812_AML1022_nanoranger_GoT.rds")

# kneeplots for exploration
knee_plot(data.table::fread("./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.GoT.ONT.csv.gz"), cutoff = 20)
ggsave("./benchmarking/figures/GoT_nanoranger/kneeplots/20230811_kneeplot_DNMT3A_GOT.ONT.png", width = 2, height = 1.5, dpi = 600)
knee_plot(data.table::fread("./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.GoT.Illumina.csv.gz"), cutoff = 20)
ggsave("./benchmarking/figures/GoT_nanoranger/kneeplots/20230811_kneeplot_DNMT3A_GOT.Illumina.png", width = 2, height = 1.5, dpi = 600)
knee_plot(data.table::fread("./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.normal.csv.gz"), cutoff = 20)
ggsave("./benchmarking/figures/GoT_nanoranger/kneeplots/20230811_kneeplot_DNMT3A_normal.png", width = 2, height = 1.5, dpi = 600)
knee_plot(data.table::fread("./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.spikein.csv.gz"), cutoff = 20)
ggsave("./benchmarking/figures/GoT_nanoranger/kneeplots/20230811_kneeplot_DNMT3A_spikein.png", width = 2, height = 1.5, dpi = 600)
knee_plot(data.table::fread("./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.spikein.additive.csv.gz"), cutoff = 20)
ggsave("./benchmarking/figures/GoT_nanoranger/kneeplots/20230811_kneeplot_DNMT3A_spikein.additive.png", width = 2, height = 1.5, dpi = 600)

knee_plot(data.table::fread("./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.GoT.ONT.csv.gz"), cutoff = 20)
ggsave("./benchmarking/figures/GoT_nanoranger/kneeplots/20230811_kneeplot_RUNX1_GOT.ONT.png", width = 2, height = 1.5, dpi = 600)
knee_plot(data.table::fread("./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.GoT.Illumina.csv.gz"), cutoff = 20)
ggsave("./benchmarking/figures/GoT_nanoranger/kneeplots/20230811_kneeplot_RUNX1_GOT.Illumina.png", width = 2, height = 1.5, dpi = 600)
knee_plot(data.table::fread("./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.normal.csv.gz"), cutoff = 20)
ggsave("./benchmarking/figures/GoT_nanoranger/kneeplots/20230811_kneeplot_RUNX1_normal.png", width = 2, height = 1.5, dpi = 600)
knee_plot(data.table::fread("./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.spikein.csv.gz"), cutoff = 20)
ggsave("./benchmarking/figures/GoT_nanoranger/kneeplots/20230811_kneeplot_RUNX1_spikein.png", width = 2, height = 1.5, dpi = 600)
knee_plot(data.table::fread("./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.spikein.additive.csv.gz"), cutoff = 20)
ggsave("./benchmarking/figures/GoT_nanoranger/kneeplots/20230811_kneeplot_RUNX1_spikein.additive.png", width = 2, height = 1.5, dpi = 600)

knee_plot(data.table::fread("./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.GoT.ONT.csv.gz"), cutoff = 20)
ggsave("./benchmarking/figures/GoT_nanoranger/kneeplots/20230811_kneeplot_SF3B1_GOT.ONT.png", width = 2, height = 1.5, dpi = 600)
knee_plot(data.table::fread("./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.GoT.Illumina.csv.gz"), cutoff = 20)
ggsave("./benchmarking/figures/GoT_nanoranger/kneeplots/20230811_kneeplot_SF3B1_GOT.Illumina.png", width = 2, height = 1.5, dpi = 600)
knee_plot(data.table::fread("./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.normal.csv.gz"), cutoff = 20)
ggsave("./benchmarking/figures/GoT_nanoranger/kneeplots/20230811_kneeplot_SF3B1_normal.png", width = 2, height = 1.5, dpi = 600)
knee_plot(data.table::fread("./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.spikein.csv.gz"), cutoff = 20)
ggsave("./benchmarking/figures/GoT_nanoranger/kneeplots/20230811_kneeplot_SF3B1_spikein.png", width = 2, height = 1.5, dpi = 600)
knee_plot(data.table::fread("./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.spikein.additive.csv.gz"), cutoff = 20)
ggsave("./benchmarking/figures/GoT_nanoranger/kneeplots/20230811_kneeplot_SF3B1_spikein.additive.png", width = 2, height = 1.5, dpi = 600)

# extract mutations for DNMT3A
DNMT3A.GoT <- extract_mutation(
  BC.data.file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.GoT.ONT.csv.gz",
  ALT = "C", REF = "T", FILTER = 4
)
DNMT3A.GoT$bc <- paste0("AML1022.1G_", DNMT3A.GoT$bc, "-1")
DNMT3A.GoT.Illumina <- extract_mutation(
  BC.data.file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.GoT.Illumina.csv.gz",
  ALT = "C", REF = "T", FILTER = 4
)
DNMT3A.GoT.Illumina$bc <- paste0("AML1022.1G_", DNMT3A.GoT.Illumina$bc)
DNMT3A.normal <- extract_mutation(
  BC.data.file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.normal.csv.gz",
  ALT = "C", REF = "T", FILTER = 4
)
DNMT3A.normal$bc <- paste0("AML1022.1_", DNMT3A.normal$bc, "-1")
DNMT3A.spikein <- extract_mutation(
  BC.data.file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.spikein.csv.gz",
  ALT = "C", REF = "T", FILTER = 4
)
DNMT3A.spikein$bc <- paste0("AML1022.1G_", DNMT3A.spikein$bc, "-1")
DNMT3A.spikein.additive <- extract_mutation(
  BC.data.file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.spikein.additive.csv.gz",
  ALT = "C", REF = "T", FILTER = 4
)
DNMT3A.spikein.additive$bc <- paste0("AML1022.1G_", DNMT3A.spikein.additive$bc, "-1")


write.table(DNMT3A.GoT,
  file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.GoT.ONT.processed.csv",
  quote = F, sep = "\t"
)
write.table(DNMT3A.GoT.Illumina,
  file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.GoT.Illumina.processed.csv",
  quote = F, sep = "\t"
)
write.table(DNMT3A.normal,
  file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.normal.processed.csv",
  quote = F, sep = "\t"
)
write.table(DNMT3A.spikein,
  file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.spikein.processed.csv",
  quote = F, sep = "\t"
)
write.table(DNMT3A.spikein.additive,
  file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.spikein.additive.processed.csv",
  quote = F, sep = "\t"
)


# extract mutations for RUNX1
RUNX1.GoT <- extract_mutation(
  BC.data.file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.GoT.ONT.csv.gz",
  ALT = "C", REF = "T", FILTER = 4
)
RUNX1.GoT$bc <- paste0("AML1022.1G_", RUNX1.GoT$bc, "-1")
RUNX1.GoT.Illumina <- extract_mutation(
  BC.data.file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.GoT.Illumina.csv.gz",
  ALT = "C", REF = "T", FILTER = 4
)
RUNX1.GoT.Illumina$bc <- paste0("AML1022.1G_", RUNX1.GoT.Illumina$bc)
RUNX1.normal <- extract_mutation(
  BC.data.file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.normal.csv.gz",
  ALT = "C", REF = "T", FILTER = 4
)
RUNX1.normal$bc <- paste0("AML1022.1_", RUNX1.normal$bc, "-1")
RUNX1.spikein <- extract_mutation(
  BC.data.file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.spikein.csv.gz",
  ALT = "C", REF = "T", FILTER = 4
)
RUNX1.spikein$bc <- paste0("AML1022.1G_", RUNX1.spikein$bc, "-1")
RUNX1.spikein.additive <- extract_mutation(
  BC.data.file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.spikein.additive.csv.gz",
  ALT = "C", REF = "T", FILTER = 4
)
RUNX1.spikein.additive$bc <- paste0("AML1022.1G_", RUNX1.spikein.additive$bc, "-1")


write.table(RUNX1.GoT,
  file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.GoT.ONT.processed.csv",
  quote = F, sep = "\t"
)
write.table(RUNX1.GoT.Illumina,
  file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.GoT.Illumina.processed.csv",
  quote = F, sep = "\t"
)
write.table(RUNX1.normal,
  file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.normal.processed.csv",
  quote = F, sep = "\t"
)
write.table(RUNX1.spikein,
  file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.spikein.processed.csv",
  quote = F, sep = "\t"
)
write.table(RUNX1.spikein.additive,
  file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.spikein.additive.processed.csv",
  quote = F, sep = "\t"
)


# extract mutations for SF3B1
SF3B1.GoT <- extract_mutation(
  BC.data.file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.GoT.ONT.csv.gz",
  ALT = "C", REF = "T", FILTER = 4
)
SF3B1.GoT$bc <- paste0("AML1022.1G_", SF3B1.GoT$bc, "-1")
SF3B1.GoT.Illumina <- extract_mutation(
  BC.data.file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.GoT.Illumina.csv.gz",
  ALT = "C", REF = "T", FILTER = 4
)
SF3B1.GoT.Illumina$bc <- paste0("AML1022.1G_", SF3B1.GoT.Illumina$bc)
SF3B1.normal <- extract_mutation(
  BC.data.file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.normal.csv.gz",
  ALT = "C", REF = "T", FILTER = 4
)
SF3B1.normal$bc <- paste0("AML1022.1_", SF3B1.normal$bc, "-1")
SF3B1.spikein <- extract_mutation(
  BC.data.file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.spikein.csv.gz",
  ALT = "C", REF = "T", FILTER = 4
)
SF3B1.spikein$bc <- paste0("AML1022.1G_", SF3B1.spikein$bc, "-1")
SF3B1.spikein.additive <- extract_mutation(
  BC.data.file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.spikein.additive.csv.gz",
  ALT = "C", REF = "T", FILTER = 4
)
SF3B1.spikein.additive$bc <- paste0("AML1022.1G_", SF3B1.spikein.additive$bc, "-1")


write.table(SF3B1.GoT,
  file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.GoT.ONT.processed.csv",
  quote = F, sep = "\t"
)
write.table(SF3B1.GoT.Illumina,
  file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.GoT.Illumina.processed.csv",
  quote = F, sep = "\t"
)
write.table(SF3B1.normal,
  file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.normal.processed.csv",
  quote = F, sep = "\t"
)
write.table(SF3B1.spikein,
  file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.spikein.processed.csv",
  quote = F, sep = "\t"
)
write.table(SF3B1.spikein.additive,
  file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.spikein.additive.processed.csv",
  quote = F, sep = "\t"
)


# generate basic statistics
DNMT3A.GoT <- read.table(file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.GoT.ONT.processed.csv")
DNMT3A.GoT.Illumina <- read.table(file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.GoT.Illumina.processed.csv")
DNMT3A.normal <- read.table(file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.normal.processed.csv")
DNMT3A.spikein <- read.table(file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.spikein.processed.csv")
DNMT3A.spikein.additive <- read.table(file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_DNMT3A.spikein.additive.processed.csv")
RUNX1.GoT <- read.table(file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.GoT.ONT.processed.csv")
RUNX1.GoT.Illumina <- read.table(file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.GoT.Illumina.processed.csv")
RUNX1.normal <- read.table(file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.normal.processed.csv")
RUNX1.spikein <- read.table(file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.spikein.processed.csv")
RUNX1.spikein.additive <- read.table(file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_RUNX1.spikein.additive.processed.csv")
SF3B1.GoT <- read.table(file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.GoT.ONT.processed.csv")
SF3B1.GoT.Illumina <- read.table(file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.GoT.Illumina.processed.csv")
SF3B1.normal <- read.table(file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.normal.processed.csv")
SF3B1.spikein <- read.table(file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.spikein.processed.csv")
SF3B1.spikein.additive <- read.table(file = "./data/benchmarking/GoT_nanoranger/AML1022_pileup_SF3B1.spikein.additive.processed.csv")

cells.1 <- length(which(AML1022$orig.ident == "AML1022.1"))
cells.2 <- length(which(AML1022$orig.ident == "AML1022.1G"))

stats <- data.frame()
stats <- rbind(stats, data.frame(
  gene = "DNMT3A",
  GoT.ONT = length(intersect(DNMT3A.GoT$bc, colnames(AML1022))), # / cells.2,
  GoT.Illumina = length(intersect(DNMT3A.GoT.Illumina$bc, colnames(AML1022))), #  / cells.2,
  normal = length(intersect(DNMT3A.normal$bc, colnames(AML1022))), #  / cells.1,
  spikein = length(intersect(DNMT3A.spikein$bc, colnames(AML1022))),
  spikein.additive = length(intersect(DNMT3A.spikein.additive$bc, colnames(AML1022)))
))
stats <- rbind(stats, data.frame(
  gene = "RUNX1",
  GoT.ONT = length(intersect(RUNX1.GoT$bc, colnames(AML1022))), # / cells.2,
  GoT.Illumina = length(intersect(RUNX1.GoT.Illumina$bc, colnames(AML1022))), # / cells.2,
  normal = length(intersect(RUNX1.normal$bc, colnames(AML1022))), # / cells.1,
  spikein = length(intersect(RUNX1.spikein$bc, colnames(AML1022))),
  spikein.additive = length(intersect(RUNX1.spikein.additive$bc, colnames(AML1022)))
))
stats <- rbind(stats, data.frame(
  gene = "SF3B1",
  GoT.ONT = length(intersect(SF3B1.GoT$bc, colnames(AML1022))), # / cells.2,
  GoT.Illumina = length(intersect(SF3B1.GoT.Illumina$bc, colnames(AML1022))), # / cells.2,
  normal = length(intersect(SF3B1.normal$bc, colnames(AML1022))), # / cells.1,
  spikein = length(intersect(SF3B1.spikein$bc, colnames(AML1022))),
  spikein.additive = length(intersect(SF3B1.spikein.additive$bc, colnames(AML1022)))
))

boo <- stats %>% tidyr::pivot_longer(names_to = "technology", cols = c("GoT.ONT", "GoT.Illumina", "normal", "spikein", "spikein.additive"))
boo$technology <- factor(boo$technology, levels = c("normal", "spikein", "GoT.Illumina", "GoT.ONT"))
ggplot(
  boo[which(boo$technology %in% c("normal", "GoT.Illumina", "GoT.ONT")), ],
  aes(x = gene, y = value, fill = technology)
) +
  geom_col(position = "dodge", color = "black") +
  scale_y_sqrt("cells genotyped", breaks = c(0, 100, 500, 1000, 2000)) +
  scale_fill_manual(values = c("GoT.ONT" = "#00aeef", "GoT.Illumina" = "black", "normal" = "orange", "spikein" = "red")) +
  theme_classic() +
  theme(
    legend.position = "none",
    axis.title = element_text("Arial", size = 10, color = "black"),
    axis.text = element_text("Arial", size = 10, color = "black"),
    axis.title.x = element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)
  )
ggsave("./benchmarking/figures/GoT_nanoranger/plots/20230912_cells_genotyped.svg", width = 2, height = 2, dpi = 600)

boo <- stats %>% tidyr::pivot_longer(names_to = "technology", cols = c("GoT.ONT", "GoT.Illumina", "normal", "spikein", "spikein.additive"))
boo$technology <- factor(boo$technology, levels = c("normal", "spikein", "spikein.additive", "GoT.Illumina", "GoT.ONT"))
ggplot(
  boo[which(boo$technology %in% c("normal", "spikein", "spikein.additive", "GoT.Illumina", "GoT.ONT")), ],
  aes(x = gene, y = value, fill = technology)
) +
  geom_col(position = "dodge", color = "black") +
  scale_y_sqrt("cells genotyped", breaks = c(0, 100, 500, 1000, 2000)) +
  scale_fill_manual(values = c(
    "GoT.ONT" = "#00aeef", "GoT.Illumina" = "black", "normal" = "orange",
    "spikein" = "red", "spikein.additive" = "firebrick"
  )) +
  theme_classic() +
  theme(
    legend.position = "none",
    axis.title = element_text("Arial", size = 10, color = "black"),
    axis.text = element_text("Arial", size = 10, color = "black"),
    axis.title.x = element_blank(),
    axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)
  )
ggsave("./benchmarking/figures/GoT_nanoranger/plots/20230920_cells_genotyped.svg", width = 2, height = 2, dpi = 600)
